src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+ * the scheduler entrypoint will deal with starting a new worker, and waiting on any txns that it needs to, if we pass along a vxid in the bgw_extra field of the BgWorker
+ *
+ */
+static bool
+register_tsbgw_entrypoint_for_db(Oid db_id, VirtualTransactionId vxid, BackgroundWorkerHandle **handle)
+{
+
+	BackgroundWorker worker;
+
+
+	memset(&worker, 0, sizeof(worker));
+	snprintf(worker.bgw_name, BGW_MAXLEN, "Timescale BGW Scheduler Entrypoint");
+	worker.bgw_flags = BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_restart_time = BGW_NEVER_RESTART;
+	worker.bgw_start_time = BgWorkerStart_RecoveryFinished;
+	sprintf(worker.bgw_library_name, EXTENSION_NAME);
+	sprintf(worker.bgw_function_name, TSBGW_ENTRYPOINT_FUNCNAME);
+	worker.bgw_notify_pid = MyProcPid;
+	worker.bgw_main_arg = db_id;
+	memcpy(worker.bgw_extra, &vxid, sizeof(VirtualTransactionId));
+
+	return RegisterDynamicBackgroundWorker(&worker, handle);
+
+}
+
+static volatile sig_atomic_t got_sigterm = false;
+static void
+tsbgw_sigterm(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+
+/*
+ * Model this on autovacuum.c -> get_database_list
+ * Note that we are not doing all the things around memory context that they do, because
+ * a) we're using shared memory to store the list of dbs and b) we're in a function and
+ * shorter lived context here.
+ * This can get called at two different times 1) when the cluster launcher starts and is looking for dbs
+ * and 2) if it restarts due to a postmaster signal.
+ */
+static HTAB *
+populate_database_htab(void)
+{
+	Relation	rel;
+	HeapScanDesc scan;
+	HeapTuple	tup;
+	HTAB	   *db_htab = NULL;
+	HASHCTL		info;
+
+	/*
+	 * first initialize the hashtable before we start a txn, we'll be writing
+	 * info about dbs there
+	 */
+	info = (HASHCTL)
+	{
+		.keysize = sizeof(Oid),
+			.entrysize = sizeof(tsbgwHashEntry)
+	};
+	db_htab = hash_create("tsbgw_launcher_db_htab", 8, &info, HASH_BLOBS | HASH_ELEM);
+
+
+	/*
+	 * by this time we should already be connected to the db, and only have
+	 * access to shared catalogs
+	 */
+	/* start a txn, see note in autovacuum.c for why */
+	StartTransactionCommand();
+	(void) GetTransactionSnapshot();
+
+	rel = heap_open(DatabaseRelationId, AccessShareLock);
+	scan = heap_beginscan_catalog(rel, 0, NULL);
+
+	while (HeapTupleIsValid(tup = heap_getnext(scan, ForwardScanDirection)))
+	{
+		Form_pg_database pgdb = (Form_pg_database) GETSTRUCT(tup);
+		tsbgwHashEntry *tsbgw_he;
+		Oid			db_oid;
+		bool		hash_found;
+
+		if (!pgdb->datallowconn)
+			continue;			/* don't bother with dbs that don't allow
+								 * connections, we'll fail when starting a
+								 * worker anyway */
+
+		db_oid = HeapTupleGetOid(tup);
+		tsbgw_he = (tsbgwHashEntry *) hash_search(db_htab, &db_oid, HASH_ENTER, &hash_found);
+		if (!hash_found)
+			tsbgw_he->db_scheduler_handle = NULL;
```

Comment:
We wanted to avoid the race condition where we have enough workers allocated to start schedulers for all databases, but before we could get all of them started, the (say) first scheduler has started too many jobs and then we don't have enough schedulers for the dbs. So we need to first get the number of dbs for which we may need schedulers, then reserve the correct number of workers, then start the workers. (will add this as a comment)